src/engine/network/network.game.ts

```diff
@@ -1,123 +1,107 @@
+  deserializeNewPeerSnapshot,
+  deserializeAssignPeerIndex,
+  deserializeSnapshot,
+  deserializeUpdatesChanged,
+  deserializeUpdatesSnapshot,
+  embodyAvatar,
+  NetPipeData,
+} from "./serialization.game";
+import { NetworkAction } from "./NetworkAction";
+import { registerInboundMessageHandler } from "./inbound.game";
+import { NetworkRingBuffer } from "./RingBuffer";
+import { deserializeCommand } from "./commands.game";
+import { InputModule } from "../input/input.game";
+import { PhysicsModule } from "../physics/physics.game";
+import { waitUntil } from "../utils/waitUntil";
 
 /*********
  * Types *
  ********/
 
 export interface GameNetworkState {
-  hosting: boolean;
-  incomingPackets: ArrayBuffer[];
-  incomingPeerIds: string[];
-  peerIdToEntityId: Map<string, number>;
-  entityIdToPeerId: Map<number, string>;
-  networkIdToEntityId: Map<number, number>;
+  incomingRingBuffer: NetworkRingBuffer<Uint8ArrayConstructor>;
+  outgoingRingBuffer: NetworkRingBuffer<Uint8ArrayConstructor>;
+  commands: ArrayBuffer[];
+  hostId: string;
   peerId: string;
   peers: string[];
   newPeers: string[];
   peerIdCount: number;
   peerIdToIndex: Map<string, number>;
+  peerIdToHistorian: Map<string, Historian>;
+  peerIdToEntityId: Map<string, number>;
+  entityIdToPeerId: Map<number, string>;
+  networkIdToEntityId: Map<number, number>;
   indexToPeerId: Map<number, string>;
   localIdCount: number;
   removedLocalIds: number[];
-  messageHandlers: { [key: number]: (input: [GameState, CursorView]) => void };
+  messageHandlers: { [key: number]: (input: NetPipeData) => void };
   cursorView: CursorView;
-  peerIdToHistorian: Map<string, Historian>;
+  // feature flags
   interpolate: boolean;
+  clientSidePrediction: boolean;
+  authoritative: boolean;
 }
 
-export enum NetworkAction {
-  Create,
-  Delete,
-  UpdateChanged,
-  UpdateSnapshot,
-  FullChanged,
-  FullSnapshot,
-  Prefab,
-  AssignPeerIdIndex,
-  InformPlayerNetworkId,
-  NewPeerSnapshot,
-  RemoveOwnershipMessage,
-}
-
-export type NetPipeData = [GameState, CursorView];
-
 /******************
  * Initialization *
  *****************/
 
 export const NetworkModule = defineModule<GameState, GameNetworkState>({
   name: "network",
-  create: (ctx): GameNetworkState => ({
-    hosting: false,
-    incomingPackets: [],
-    incomingPeerIds: [],
-    networkIdToEntityId: new Map<number, number>(),
-    peerIdToEntityId: new Map(),
-    entityIdToPeerId: new Map(),
-    peerId: "",
-    peers: [],
-    newPeers: [],
-    peerIdToIndex: new Map(),
-    indexToPeerId: new Map(),
-    peerIdCount: 0,
-    localIdCount: 0,
-    removedLocalIds: [],
-    messageHandlers: {},
-    cursorView: createCursorView(),
-    peerIdToHistorian: new Map(),
-    interpolate: false,
-  }),
+  create: async (ctx, { waitForMessage }): Promise<GameNetworkState> => {
+    const { incomingRingBuffer, outgoingRingBuffer } = await waitForMessage<InitializeNetworkStateMessage>(
```

Comment:
Lets expose the feature flags from a query string and send them over to the game thread here.

Reply:
it would probably be wise to add explicit `setup` and `teardown` hooks to the modules which can be induced when joining/leaving a room